### PR TITLE
Add support for adding to the environment from status-react

### PR DESCRIPTION
**Pull request by oskarth**
Add support for adding to the environment from status-react.

Adds AddEnv public API call that enables status-react (and by extensions, Jenkins CI for QA) to set env variable values in status-go.

This is useful for feature flags, where we just want to instrument a very specific part of the code without error-prone changes to the config.

For status-go developers, these ENV variables work the same as normal ENV lookups when running in more forgiving environment than bundled as an ObjC package for a React Native app. The interface is thus string based, just like env variables set in shell.

Important changes:
- [x] AddEnv(key, value) API call.

Solves the problem of enabling:
- https://github.com/status-im/status-react/issues/2544
- https://github.com/status-im/status-react/issues/2504
- https://github.com/status-im/status-react/issues/2511
- https://github.com/status-im/status-go/pull/464
- https://github.com/status-im/status-go/issues/432

status: ready

### CODE REVIEW

File: geth/api/api.go [hunk 1]
```diff
@@ -234,3 +235,15 @@ func (api *StatusAPI) NotifyUsers(message string, payload fcm.NotificationPayloa
 
 	return err
 }
+
+// AddEnv loads key-value pairs into ENV, such as flags.
+func (api *StatusAPI) AddEnv(key string, value string) string {
```
Comment (adambabik) [inline geth/api/api.go:240]: If it calls `os.Setenv`, I would call this binding `SetEnv`.

File: geth/api/api.go [hunk 2]
```diff
@@ -234,3 +235,15 @@ func (api *StatusAPI) NotifyUsers(message string, payload fcm.NotificationPayloa
...
+	log.Debug("AddEnv", key, value)
+
+	err := os.Setenv(key, value)
+	if err != nil {
+		log.Error("AddEnv error:", err)
+	}
+
+	return key
```
Comment (adambabik) [inline geth/api/api.go:248]: I guess no need to return key. As far as I know, we usually do `{"error":""}` if there are no errors (which is not ideal but at least we will keep it consistent).

File: lib/library.go [hunk 3]
```diff
@@ -442,3 +442,9 @@ func NotifyUsers(message, payloadJSON, tokensArray *C.char) (outCBytes *C.char)
 
 	return
 }
+
+//export AddEnv
+func AddEnv(key *C.char, value *C.char) *C.char {
```
Comment (adambabik) [inline lib/library.go:447]: `s/AddEnv/SetEnv` maybe?

**Comment by divan** [comment]
I'm not really convinced that environment vars are the good choice for configuring `status-go` behavior and the API call `SetEnv` looks like a workaround for "we can't set env variables in our environment" problem.

Anyway, here is my explanation why I don't think env vars are a proper way to do configuration. That's what we have currently — all configuration is passed through config file, either read from the file, or generated by `status-react`:

```  
    +------------------------------------+
    |            Environment             |
    |------------------------------------|
    |+------------+      +--------------+|
    ||Status-react|      |Status-go     ||
    ||------------|      |--------------||
    ||            | API  |              ||
    ||            <------+              ||
    ||            +------>              ||
    ||            |      |              ||
    ||            |      +--------------+|
    ||            +------>Config        ||
    |+------------+      +--------------+|
    +------------------------------------+
```

Config file approach has its own drawbacks, but that's how `ethereum-go` is configured and that's how `status-go` was designed. It also aligns with best practice to keep configuration in one place if possible and pass parts of it down the stack if needed (see `WhisperConfig`, for example). 

We don't use env vars currently, except a few very specific cases where we really requesting information about the environment (`GOPATH` for printing build info, for example).

---

Mentioned issues (all related to https://github.com/status-im/status-go/issues/432 in fact) do not need env vars. A decision to use env var in https://github.com/status-im/status-go/pull/464 was meant to be the quickest way to add a temporary knob, which never meant to stay in production code — idea was to make custom build, collect data and then implement a proper solution based on this data.

This PR embraces this "env vars" approach and adds new API call to set env variable, which is in fact subverting the idea of "environment vars". If value is set via API call it's not an environment variable, it's a valid configuration value.

So passing configuration value as an environmental variable through API call (instead of reading from the environment) seems to me like a very confusing solution. Even more, if we introduce a second way to configure behavior, it'll make developers wonder which one to use — config file or vars. It also will make testing harder — now we'll need to make sure that env variables are initialized with proper values along with configuration and we also lose visibility of what configuration values are even available.

Hence I recommend sticking only to one way of passing configuration to status-go.

**Comment by oskarth** [comment]
To clarify, I did try the configuration approach end to end (RCTStatus.m and a lot of different files in status-go, similar to devMode/networkID). It smelt funny and requires a lot of changes, some of which I couldn't even get to work in a reasonable amount of time due to needing to update config stuff in a dozen places.

I'm going afk for a while now (will try to respond more thoroughly later to your thoughtful comment above), but can you please illustrate an alternative solution for how you'd go to:
- An if statement for a behavior being just an if statement and maybe one place to 'add' the toggle (flag/env/cfg) as a thing, without 20 line error-prone diff
- No changes in RCTStatus.m required to effect this change

As to the specific implementation detail of using `os.SetEnv`, that of course doesn't matter as long as it aids the solution the interface `AddEnv` (can be renamed to anything) provides.

**Review by adambabik — APPROVED** [review]
In case a status-go experiment requires env to be set beforehand, I guess it will still be possible, right? Is it not possible to start an iOS app with some env set in Xcode for example?

**Comment by divan** [comment]
@oskarth 

 > but can you please illustrate an alternative solution for how you'd go to:

So far we had only one case where we need this, and it's perfectly solvable by using env vars (and code in this PR) in a separate branch, without introducing this functionality into `develop`.

Let's clarify what this PR aims to do — introduce an easy way to have feature flags, right? That's what is usually called `Feature Toggles`. API call could be `ToggleFeature()` then instead of `AddEnv()`. But I'm questioning not only implementation (again, I don't want to have codebase peppered with `os.Getenv()` every now and then for configuration), but the idea of introducing Feature Toggle functionality itself.

Without any doubt, feature toggles could be useful, but they also introduce another layer of complexity and technical debt to deal with:
 - exact implementation (this discussion about os envs, for example)
 - naming (need to keep track of available feature toggles, maintain docs etc)
 - visibility of toggles (I had issues with software deployed and toggled flag accidentally without anyone being aware of that more than once)
 - control/report access to who can enable flags
 - remove old flags (it's more complicated than cleaning dead code and usually is never done)
 - etc

Bottom line, it's not just "add another API call and solve the problem of feature toggle" it's the whole new functionality that needs constant investments of time and efforts, needs to be supported and maintained.

So what I'm trying to ask here — do we really need to introduce Feature Toggles into status-go? We have one case where we are totally fine with a custom build from a temporary branch.

If the use case for Feature Toggles will appear again, let's consider it then. For now, I believe, it adds more complexity then benefits.

**Review by tiabc — CHANGES_REQUESTED** [review]
Agree with @divan

**Comment by adambabik** [comment]
Is the issue resolved without this code? Can we close this PR?

**Comment by oskarth** [comment]
@adambabik IMO no. Can we leave it open for a bit longer? had on my TODO to reply but haven't prioritized it yet.

**Comment by oskarth** [comment]
Closing this one as it is in 'solution space' (PRs) instead of 'problem space' (issues), and we disagree on whether this is a problem or not. Will likely re-open an issue of this at some point and link to this PR then.
